### PR TITLE
Fix updating records when editing a submission

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Ensures that when editing an existing submission, the previously submitted contacts, etc will be updated.

Before
----------------------------------------

**Problem 1**
1. Make a "create new contact" style webform (without existing contact matching)
2. Submit the form. A new contact will be created.
3. Go back and edit the submission, then submit again.
4. Instead of updating the contact previously created, it creates a 2nd contact.

**Problem 2**
1. Add a multi-record custom group which contains a text field and a contact ref field
2. Expose both to a "create new contact" style webform, adding a 2nd contact to the form as the value of the contact reference field.
3. When saving the form, 2 records will be inserted, one for each field. Should have been a single record with both values.

**Problem 3**
1. Add a multi-record custom group with some fields
2. Make an "update current user" style webform (existing contact match to user)
3. Add one set of the multi-value fields, selecting "Create Only" for the create mode.
4. Submit the form. Note one custom record has been inserted.
5. Edit the form submission and save. Note a second record has been inserted instead of updating the previously created one.

After
----------------------------------------
All the above work as expected now.

### CODE REVIEW

**Comment by colemanw** [comment]
@KarinG I'm impressed at the existing test coverage! Tests caught several bugs in this PR, and after several rounds of fixes I got everything passing, and then added a unit test of my own (my very first :heart_eyes:). It tests the scenario described in problem 1, which I think is the most important. It would be nice to have tests for the other two but IMO shouldn't block merging this.

**Comment by KarinG** [comment]
@colemanw - awesome on your first test 🎉 I love these functional tests - they are meaningful and easy to understand and fun to write. And have been able to identify regressions in CiviCRM (and recently in Webform). I’ll be reviewing this on Monday.

**Comment by KarinG** [comment]
PS and yes I don’t have a must include tests for all scenarios policy so no worries.

**Comment by KarinG** [comment]
Sorry @colemanw - this needs a rebase. 

PS - on testing - have you discovered the browser artifacts yet and the screenshots?

**Comment by jitendrapurohit** [comment]
Thanks @colemanw. Saving/updating multi-record values in create/edit mode had some complexities and thanks for cleaning those up 👍 .  I'm glad the tests are passing with all these changes.

I've tested with 3-5 multi-valued custom fields (including contact refs which have bitten us in the past) and confirmed that the values are stored/updated as expected.

**Comment by colemanw** [comment]
Thanks for the review @jitendrapurohit - I've rebased and tests are still passing so I think this is good to go :)

**Comment by colemanw** [comment]
@KarinG no, what are the browser artifacts and screenshots?

**Comment by KarinG** [comment]
Hang on there! I’m still going to review this tomorrow. And run my usual live site checks. And then I’ll show you what the artifacts and screenshots are as well.

**Comment by KarinG** [comment]
Existing automated tests ✅
New test - local check ✅

Live site checks
Problem 1 (most important one) ✅
Screenshots

Edit submission -> add Middle Name -> Dog 
image

Still only one record in CiviCRM ✅ and also checked that contact record is updated to include Middle Name (Dog)
image

**Comment by KarinG** [comment]
Browser artifacts: located at the bottom of this screen -> currently hundreds of HTML files + a handful of screenshots. 

image

HTML from your new test:
image

image

html is generated by the test on certain events e.g. when a button is hit - but you can always request one by injecting:

`$this->htmlOutput();`

e.g. https://github.com/colemanw/webform_civicrm/blob/6.x/tests/src/FunctionalJavascript/ContactSubmissionTest.php#L212

In addition there are screenshots PNG -> e.g. -> alaxvalues.png -> 

`$this->createScreenshot($this->htmlOutputDirectory . '/ajaxvalues.png');`

https://github.com/colemanw/webform_civicrm/blob/6.x/tests/src/FunctionalJavascript/CustomFieldSubmissionTest.php#L553

image

**Comment by KarinG** [comment]
So that 👆 is a screenshot generated by Github Actions -> currently only the last pass (of the 4 Drupal x CiviCRM combinations in our testing matrix; which CiviCRM: see title of the webform -> 5.39.1 in this case) - it's super helpful to debug tests and see what's going on - especially if there is a mystery difference between your local and Github actions.

I also use screenshots on my local when writing tests. 

At some point I'll cull the html and screenshots a bit when we need to start speeding up the tests - but right now the tests are ok in terms of length and I think these visuals are an important part of testing.

**Comment by KarinG** [comment]
Forgot to say THANK you for jumping in and getting this piece ported to D9. It is much appreciated. Hope you enjoyed it!

**Comment by colemanw** [comment]
Thanks @KarinG for the thorough revew & explanation!

File: tests/src/FunctionalJavascript/SubmissionEditTest.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    $this->assertEquals('Smarty', civicrm_api3('Contact', 'get', ['id' => $newMax])['values'][$newMax]['first_name']);
+
+  }
+
+  protected function getMaxId($entity = 'Contact') {
+    return civicrm_api3($entity, 'get', [
+      'options' => ['sort' => "id DESC", 'limit' => 1]
+    ])['id'];
+  }
```
Comment (colemanw) [inline tests/src/FunctionalJavascript/SubmissionEditTest.php:64]: FYI @KarinG in an earlier draft of this PR I was using APIv4 here. The test worked locally, but bombed on GitHub.
Ah well, someday we'll move this module up to v4.

Interesting historical note, APIv3 and this module are very nearly the same age; version 1.0 of Webform-CiviCRM was released just a month or so after APIv3 was introduced in CiviCRM core.

Comment (KarinG) [reply tests/src/FunctionalJavascript/SubmissionEditTest.php:64]: @colemanw - the plan for that is to leave the tests as is (APIv3 - for now) - but to start with replacing all APIv3 calls with APIv4 calls in the code itself - first. I believe we already have 1 or 2 🙂

**Pull request merged**